crates/bevy_pbr/src/light_probe/generate.rs

```diff
@@ -0,0 +1,1084 @@
+//! Generated environment map filtering.
+//!
+//! A *generated environment map* converts a single, high-resolution cubemap
+//! into the pair of diffuse and specular cubemaps required by the PBR
+//! shader. Add [`bevy_light::GeneratedEnvironmentMapLight`] to a camera
+//! and Bevy will, each frame, generate the diffuse and specular cubemaps
+//! required by the PBR shader.
+//!
+//! 1. Copy the base mip (level 0) of the source cubemap into an intermediate
+//!    storage texture.
+//! 2. Generate mipmaps using single-pass down-sampling (SPD).
+//! 3. Convolve the mip chain twice:
+//!    * a Lambertian convolution for the 32 × 32 diffuse cubemap
+//!    * a GGX convolution, once per mip level, for the specular cubemap.
+//!
+//! The filtered results are then consumed exactly like the textures supplied
+//! by [`bevy_light::EnvironmentMapLight`]. This is useful when you only have a
+//! raw HDR environment map or when you need reflections generated at run time.
+//!
+//! [single-pass down-sampling]: <SPD-paper-URL>
+//! [Lambertian convolution]: <reference-URL>
+//! [GGX convolution]: <reference-URL>
+use bevy_asset::{load_embedded_asset, uuid_handle, AssetServer, Assets, Handle};
+use bevy_ecs::{
+    component::Component,
+    entity::Entity,
+    query::{QueryState, With, Without},
+    resource::Resource,
+    system::{lifetimeless::Read, Commands, Query, Res, ResMut},
+    world::{FromWorld, World},
+};
+use bevy_image::Image;
+use bevy_math::{Quat, Vec2};
+use bevy_render::{
+    render_asset::{RenderAssetUsages, RenderAssets},
+    render_graph::{Node, NodeRunError, RenderGraphContext, RenderLabel},
+    render_resource::{
+        binding_types::*, AddressMode, BindGroup, BindGroupEntries, BindGroupLayout,
+        BindGroupLayoutEntries, CachedComputePipelineId, ComputePassDescriptor,
+        ComputePipelineDescriptor, Extent3d, FilterMode, PipelineCache, Sampler,
+        SamplerBindingType, SamplerDescriptor, ShaderDefVal, ShaderStages, ShaderType,
+        StorageTextureAccess, Texture, TextureAspect, TextureDescriptor, TextureDimension,
+        TextureFormat, TextureSampleType, TextureUsages, TextureView, TextureViewDescriptor,
+        TextureViewDimension, UniformBuffer,
+    },
+    renderer::{RenderContext, RenderDevice, RenderQueue},
+    settings::WgpuFeatures,
+    sync_world::RenderEntity,
+    texture::{CachedTexture, GpuImage, TextureCache},
+    Extract,
+};
+
+use bevy_light::{EnvironmentMapLight, GeneratedEnvironmentMapLight};
+use core::cmp::min;
+
+/// Handle for Spatio-Temporal Blue Noise texture
+pub const STBN: Handle<Image> = uuid_handle!("3110b545-78e0-48fc-b86e-8bc0ea50fc67");
+
+/// Labels for the environment map generation nodes
+#[derive(PartialEq, Eq, Debug, Copy, Clone, Hash, RenderLabel)]
+pub enum GeneratorNode {
```

Comment:
Imo, don't make separate nodes for each step. I'd much rather a single node, unless you have a strong reason to separate it.

Reply:
i merged two into `FilteringNode`, leaving only two render-graph nodes overall: `DownsamplingNode` and `FilteringNode` The downsample pass is still in its own node because those mips are broadly useful elsewhere, exposure, or down-sized probes, texture LODs. Keeping that one node separate lets us plug its output into other graph branches without rerunning the expensive convolution. Everything else now runs in one compute pass block, so per-frame graph overhead is lower.